File: cdp-agentkit-core/typescript/src/actions/cdp/data/pyth/fetch_price_feed_id.ts
import { CdpAction } from "../../cdp_action";
import { z } from "zod";

const PYTH_FETCH_PRICE_FEED_ID_PROMPT = `
Fetch the price feed ID for a given token symbol from Pyth.
`;

/**
 * Input schema for Pyth fetch price feed ID action.
 */
export const PythFetchPriceFeedIDInput = z.object({
  tokenSymbol: z.string().describe("The token symbol to fetch the price feed ID for"),
});

/**
 * Fetches the price feed ID from Pyth given a ticker symbol.
 *
 * @param args - The input arguments for the action.
 * @returns A message containing the price feed ID corresponding to the given ticker symbol.
 */
export async function pythFetchPriceFeedID(
  args: z.infer<typeof PythFetchPriceFeedIDInput>,
): Promise<string> {
  const url = `https://hermes.pyth.network/v2/price_feeds?query=${args.tokenSymbol}&asset_type=crypto`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const data = await response.json();

  if (data.length === 0) {
    throw new Error(`No price feed found for ${args.tokenSymbol}`);
  }

  const filteredData = data.filter(
    (item: any) => item.attributes.base.toLowerCase() === args.tokenSymbol.toLowerCase(),
  );

  if (filteredData.length === 0) {
    throw new Error(`No price feed found for ${args.tokenSymbol}`);
  }

  return filteredData[0].id;
}

/**
 * Pyth fetch price feed ID action.
 */
export class PythFetchPriceFeedIDAction implements CdpAction<typeof PythFetchPriceFeedIDInput> {
  public name = "pyth_fetch_price_feed_id";
  public description = PYTH_FETCH_PRICE_FEED_ID_PROMPT;
  public argsSchema = PythFetchPriceFeedIDInput;
  public func = pythFetchPriceFeedID;
}

File: python/cdp-agentkit-core/cdp_agentkit_core/actions/pyth/fetch_price_feed_id.py
from collections.abc import Callable

import requests
from pydantic import BaseModel, Field

from cdp_agentkit_core.actions import CdpAction

PYTH_FETCH_PRICE_FEED_ID_PROMPT = """
Fetch the price feed ID for a given token symbol (e.g. BTC, ETH, etc.) from Pyth.
"""


class PythFetchPriceFeedIDInput(BaseModel):
    """Input schema for fetching Pyth price feed ID."""

    token_symbol: str = Field(..., description="The token symbol to fetch the price feed ID for.")


def pyth_fetch_price_feed_id(token_symbol: str) -> str:
    """Fetch the price feed ID for a given token symbol from Pyth."""
    url = f"https://hermes.pyth.network/v2/price_feeds?query={token_symbol}&asset_type=crypto"
    response = requests.get(url)
    response.raise_for_status()
    data = response.json()

    if not data:
        raise ValueError(f"No price feed found for {token_symbol}")

    filtered_data = [
        item for item in data if item["attributes"]["base"].lower() == token_symbol.lower()
    ]
    if not filtered_data:
        raise ValueError(f"No price feed found for {token_symbol}")

    return filtered_data[0]["id"]


class PythFetchPriceFeedIDAction(CdpAction):
    """Pyth Fetch Price Feed ID action."""

    name: str = "pyth_fetch_price_feed_id"
    description: str = PYTH_FETCH_PRICE_FEED_ID_PROMPT
    args_schema: type[BaseModel] | None = PythFetchPriceFeedIDInput
    func: Callable[..., str] = pyth_fetch_price_feed_id
